### PR TITLE
Convert backend to TypeScript (84% -> 89%)

### PR DESCRIPTION
**Pull request by eliandoran**
Language|files|blank|comment|code
:-------|-------:|-------:|-------:|-------:
TypeScript|217|5587|1618|21490
JavaScript|33|653|92|2619
--------|--------|--------|--------|--------
SUM:|250|6240|1710|24109

### CODE REVIEW

**Comment by ProfDoof** [comment]
Hey, I had a realization, you're in a perfect spot to do a short blurb level of documentation about the overall architecture with the rewrite. If you have time, would you mind doing that? Maybe with some diagrams in mermaid.js or something as well? That would be really helpful to make contributions even easier later.

**Comment by eliandoran** [comment]
> Hey, I had a realization, you're in a perfect spot to do a short blurb level of documentation about the overall architecture with the rewrite. If you have time, would you mind doing that? Maybe with some diagrams in mermaid.js or something as well? That would be really helpful to make contributions even easier later.

@ProfDoof , I think it's a good idea but I wouldn't say I'm yet proficient with the code for the server. The main reason is that during conversion I mostly focused on solving build errors, so there's still a lot of digging to do before I can come up with some useful documentation.

Nevertheless, I take the point and start taking some notes as I grow more accustomed to the application.

File: src/share/shaca/entities/snote.ts [hunk 1]
```diff
@@ -1,70 +1,71 @@
...
-    /** @returns {SBranch[]} */
     getBranches() {
         return this.parentBranches;
     }
 
-    /** @returns {SBranch[]} */
-    getChildBranches() {
+    getChildBranches(): SBranch[] {
         return this.children.map(childNote => this.shaca.getBranchFromChildAndParent(childNote.noteId, this.noteId));
     }
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/share/shaca/entities/snote.ts:71]: @alexpietsch , I used those to help me when there were a lot of cross-dependencies between the different types of entities, and forgot to remove some of them afterwards. Did it retroactively for all the redundant type comments.

File: src/share/shaca/entities/snote.ts [hunk 2]
```diff
@@ -125,43 +126,41 @@ class SNote extends AbstractShacaEntity {
...
-     * @returns {SAttribute[]} all note's attributes, including inherited ones
+     * @param type - (optional) attribute type to filter
+     * @param name - (optional) attribute name to filter
+     * @returns all note's attributes, including inherited ones
      */
-    getAttributes(type, name) {
-        if (!this.__attributeCache) {
-            this.__getAttributes([]);
+    getAttributes(type?: string, name?: string) {
+        let attributeCache = this.__attributeCache;
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/share/shaca/entities/snote.ts:140]: @alexpietsch , yes, this method made heavy use of `this._attributeCache` which is nullable, so I extracted it as a local variable instead in order not to have to add extra null checks.

File: src/share/shaca/entities/snote.ts [hunk 3]
```diff
@@ -212,214 +211,235 @@ class SNote extends AbstractShacaEntity {
...
-    /** @returns {SAttribute[]} */
-    __getInheritableAttributes(path) {
+    __getInheritableAttributes(path: string[]) {
         if (path.includes(this.noteId)) {
             return [];
         }
 
         if (!this.__inheritableAttributeCache) {
-            this.__getAttributes(path); // will refresh also this.__inheritableAttributeCache
+            return this.__getAttributes(path); // will refresh also this.__inheritableAttributeCache
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/share/shaca/entities/snote.ts:220]: @alexpietsch , another very good catch! :)

File: src/share/shaca/entities/snote.ts [hunk 4]
```diff
@@ -212,214 +211,235 @@ class SNote extends AbstractShacaEntity {
...
+    hasOwnedRelation(name: string) { return this.hasOwnedAttribute(RELATION, name); }
 
     /**
-     * @param {string} name - label name
-     * @returns {SAttribute|null} label if it exists, null otherwise
+     * @param name - label name
+     * @returns label if it exists, null otherwise
      */
-    getLabel(name) { return this.getAttribute(LABEL, name); }
+    getLabel(name: string) { return this.getAttribute(LABEL, name); }
```
[the remark this answers is not in the document]
Comment (eliandoran) [reply src/share/shaca/entities/snote.ts:302]: Seems OK now.

**Comment by eliandoran** [comment]
@alexpietsch , please re-review.

**Pull request merged**